### PR TITLE
Update laminas-code to ^3.5.1 to be PHP8 compatible

### PR DESCRIPTION
**Pull request by hws47a**
### Description (*)

Update laminas-code to ^3.5.1 to be PHP8 compatible

### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/513

### Fixed Issues (if relevant)

1. Fixes magento/magento2#31345

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by hws47a** [comment]
@magento run all tests

**Comment by hws47a** [comment]
@magento run all tests

**Comment by hws47a** [comment]
In Static Tests:
```
1) Warning
The data provider specified for Magento\Test\Integrity\PublicCodeTest::testAllPHPClassesReferencedFromPublicClassesArePublic is invalid.
PHPUnit\Framework\Exception: Warning: Use of undefined constant T_NAME_FULLY_QUALIFIED - assumed 'T_NAME_FULLY_QUALIFIED' (this will throw an Error in a future version of PHP) in /var/www/html/vendor/laminas/laminas-code/src/Scanner/ClassScanner.php:1002.
/var/www/html/dev/tests/static/framework/bootstrap.php:62
/var/www/html/vendor/laminas/laminas-code/src/Scanner/ClassScanner.php:1002
/var/www/html/vendor/laminas/laminas-code/src/Scanner/ClassScanner.php:190
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/PublicCodeTest.php:152
```
This is the issue with backwards compatibility in laminas-code v3.5.1. I'll create a PR to their repo a bit later and hope it'll be fixed.

They say that in 3.5.1 they cannot guarantee that it'll work in PHP 8.0 while in 4.0.0 they dropped support of PHP 7.3

**Comment by ihor-sviziev** [comment]
@hws47a if that constant was removed - maybe it was just moved somewhere? If so - we can use our own private constant

**Comment by hws47a** [comment]
I've created a PR for laminas-code 3.5.x branch. Let's see how it'll go and if slowly we may temporary define constants in Magento

**Comment by hws47a** [comment]
@magento run all tests

**Comment by hws47a** [comment]
@magento run all tests

**Comment by hws47a** [comment]
@magento run all tests

**Comment by hws47a** [comment]
@magento run all tests

**Comment by hws47a** [comment]
@magento run Unit Tests

**Comment by hws47a** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please resolve conflicts. Seems like that's causing test failures

**Comment by hws47a** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileScanner.php [hunk 1]
```diff
@@ -38,6 +38,14 @@ protected function scan()
             define('T_TRAIT', 42001);
         }
 
+        // ensure php backwards compatibility (from laminas code 3.5.x)
+        if (! defined('T_NAME_QUALIFIED')) {
+            define('T_NAME_QUALIFIED', 24001);
+        }
+        if (! defined('T_NAME_FULLY_QUALIFIED')) {
+            define('T_NAME_FULLY_QUALIFIED', 24002);
+        }
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/FileScanner.php]: Let's format all if statements in single format - space isn't needed here.
```suggestion
        if (!defined('T_NAME_QUALIFIED')) {
            define('T_NAME_QUALIFIED', 24001);
        }
        if (!defined('T_NAME_FULLY_QUALIFIED')) {
            define('T_NAME_FULLY_QUALIFIED', 24002);
        }
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Could you fix one small issue and squash your changes to single commit?

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileScanner.php (hunk 1 above)
Comment (hws47a) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/FileScanner.php]: @ihor-sviziev, it’s not ready for review yet. I try to mark it as “dev in progress” every time.

Comment (ihor-sviziev) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/FileScanner.php]: Ah, that's label for issues. The correct label is "needs update". Also you can convert PR to draft

Comment (hws47a) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/FileScanner.php]: Thank you, will do the Draft next time.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by hws47a** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Comment by hws47a** [comment]
@ihor-sviziev all checks finally passed. Thank you

**Review by ihor-sviziev — APPROVED** [review]

**Pull request merged**